### PR TITLE
Limit latest-post authors dropdown to users with published posts

### LINKED ISSUE #32620: Unfiltered user REST API requests
## Description
There are two places where Gutenberg sends unfiltered requests for literally ALL users.

On large sites this can be hundreds of thousands of users and thus causes a noticeable load on your server for seemingly no use at all.

These happen in the latest posts block:
https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/latest-posts/edit.js#L53

And in the user autocompleter:
https://github.com/WordPress/gutenberg/blob/trunk/packages/editor/src/components/autocompleters/user.js#L22

I can think of no reason why either of these would need to fetch literally every single user, which is what the REST controller does if the user making the request has the `list_users` permission.

## Step-by-step reproduction instructions
1. Create 10.000 users.
2. Be logged in as an admin.
3. Insert a latest posts block.
## Expected behaviour

I'd expect both these requests to be done with the `who=authors` parameter or alternatively the `getAuthors` selector.

## Actual behaviour
These requests keep on paginating for potentially hundreds of pages fetching every single user.

### PR DESCRIPTION
**Pull request by aristath**
## Description
* Adds a `has_published_posts` argument for the user query in the users REST API.
* Limits the latest-post block's "author" dropdown to only users with published posts.
 
Fixes #32620 

## How has this been tested?
In a test-site with 10k users, adding a latest-posts block was causing significant server load (see description in #32620). After this PR, the request is almost instant and only shows the 5 users that have published posts on my localhost instead of a dropdown with 10k users.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

File: lib/rest-api.php
Comment (oandregal) [inline lib/rest-api.php]:
```suggestion
```

Comment (aristath) [reply lib/rest-api.php]: :laughing: yeah, I missed this in the initial commit. Pushed a fix for this already in https://github.com/WordPress/gutenberg/pull/32662/commits/9b33ffbde33c9cff5b5e4c7dfe761055caa713cf :+1:

File: lib/rest-api.php
Comment (oandregal) [inline lib/rest-api.php]: What would be the difference from using `$prepared_args['has_published_posts']=true`? Upon [reading the docs](https://developer.wordpress.org/reference/classes/wp_user_query/#has-published-posts-parameter), I'm not sure whether there're post types that aren't shown in rest requests but can be public.

File: packages/block-library/src/latest-posts/edit.js
Comment (oandregal) [inline packages/block-library/src/latest-posts/edit.js]: I don't have enough exposure to the REST API to tell whether making the parameters work differently than the `WP_User_Query` is something we allow to do, so I'll leave others to comment on this aspect.

For my own education, I wanted to gather your thoughts on an alternative: what if we added support for `has_published_post`? My understanding is that in the case of the latest-posts block it could be enough if we passed `true` to the endpoint:

```js
const USERS_LIST_QUERY = {
  per_page: -1,
  has_published_posts: true,
};
```

File: lib/rest-api.php
Comment (aristath) [reply lib/rest-api.php]: I don't think it will make any difference, and I'd probably just use `true` as well.
The reason I did it using `get_post_types( array( 'show_in_rest' => true ), 'names' );` is that I was thinking of making it easier to backport this change to core later on. So this is just for consistency with https://github.com/WordPress/wordpress-develop/blob/5cd6bfe2e0f35922c70b4243288d49f239a09ab4/src/wp-includes/rest-api/endpoints/class-wp-rest-users-controller.php#L301 and backporting would simply require changing the condition above that line.
We can change it to `true`, and I can add a note in the function comments that this should also be changed in core when backporting if you think that's best :+1:

File: packages/block-library/src/latest-posts/edit.js
Comment (aristath) [reply packages/block-library/src/latest-posts/edit.js]: That would also work... I was simply trying to make the changes minimally invasive, and the `who` arg seemed like a reasonable way to accomplish what we need. I'll see if I can make that work :+1:

Comment (aristath) [reply packages/block-library/src/latest-posts/edit.js]: I think your suggestion makes more sense than what I was initially doing... Pushed a fix and now it has a `has_published_posts (boolean)` arg :+1:

File: lib/rest-api.php
Comment (oandregal) [reply lib/rest-api.php]: Nice, it makes sense leaving it this way then.

**Comment by TimothyBJacobs** [comment]
Can this use the `getAuthors` selector https://github.com/WordPress/gutenberg/blob/trunk/packages/core-data/src/selectors.js#L60 and behave like the post author dropdown?

File: lib/rest-api.php
Comment (oandregal) [inline lib/rest-api.php]: This needs to be updated with the new argument.

**Comment by aristath** [comment]
> Can this use the getAuthors selector https://github.com/WordPress/gutenberg/blob/trunk/packages/core-data/src/selectors.js#L60 and behave like the post author dropdown?

It can... but I'm not sure it should. As mentioned in the ticket's [comment](https://github.com/WordPress/gutenberg/issues/32620#issuecomment-859642832):

> Adding `who=authors` may not be suitable... non-authors may be able to create posts for other post-types, or they may have authored 100 posts before being "demoted" and are no longer an author - in which case they should still be in the list :thinking:

File: lib/rest-api.php
Comment (oandregal) [inline lib/rest-api.php]: The `WP_Users_Query`'s `has_published_posts` param can be of type boolean or array. I understand we may want to do the same in the REST API. Looked a bit for examples of similar cases but couldn't find any. Are you aware what are our options here?

**Comment by TimothyBJacobs** [comment]
> > Can this use the getAuthors selector https://github.com/WordPress/gutenberg/blob/trunk/packages/core-data/src/selectors.js#L60 and behave like the post author dropdown?
> 
> It can... but I'm not sure it should. As mentioned in the ticket's [comment](https://github.com/WordPress/gutenberg/issues/32620#issuecomment-859642832):
> 
> > Adding `who=authors` may not be suitable... non-authors may be able to create posts for other post-types, or they may have authored 100 posts before being "demoted" and are no longer an author - in which case they should still be in the list 🤔

That seems sensible. We need to add unit tests for the new parameter then.

File: lib/rest-api.php
Comment (TimothyBJacobs) [reply lib/rest-api.php]: Given how close we are to release, let's just stick with a boolean for now.

Comment (aristath) [reply lib/rest-api.php]: Done in https://github.com/WordPress/gutenberg/pull/32662/commits/09dce10af7bc70ea94f722a9ff07b39416bee45d, and also changed the latest-posts block's `has_published_posts` to `[ 'post' ]` in https://github.com/WordPress/gutenberg/pull/32662/commits/c54306d11098b9529e24f74a85565e3b01e2ce85 'cause we don't care for other post-types there :+1:

**Comment by aristath** [comment]
> We need to add unit tests for the new parameter then.

Done in https://github.com/WordPress/gutenberg/pull/32662/commits/fed7009d71188c8d7b728377d16bf4e76d5fdbc0

Note: Most of that commit is just copied from wp-core, the only things added (which would require backporting to core) are [line 180](https://github.com/WordPress/gutenberg/pull/32662/commits/fed7009d71188c8d7b728377d16bf4e76d5fdbc0#diff-c9eed934f81631a5b70cf2202e0ec2a804ab08ba6a9f5c402ec277f3ee24fe85R180) and [lines 196-222](https://github.com/WordPress/gutenberg/pull/32662/commits/fed7009d71188c8d7b728377d16bf4e76d5fdbc0#diff-c9eed934f81631a5b70cf2202e0ec2a804ab08ba6a9f5c402ec277f3ee24fe85R196-R222)

**Review by oandregal — APPROVED** [review]
It works as advertised. The code looks good as well.

Before merging, I'd like to hear a second opinion by folks with more mileage than myself on REST API matters ― specially on how the new argument has declared [two potential types](https://github.com/WordPress/gutenberg/pull/32662/files#diff-d0fe708f50f3e61504855aaa9f076b0c8ad1c07c185e7e18f68fcdd22c99ed21R218), as I haven't found any other example that does the same.

**Comment by aristath** [comment]
> Before merging, I'd like to hear a second opinion by folks with more mileage than myself on REST API matters ― specially on how the new argument has declared [two potential types](https://github.com/WordPress/gutenberg/pull/32662/files#diff-d0fe708f50f3e61504855aaa9f076b0c8ad1c07c185e7e18f68fcdd22c99ed21R218), as I haven't found any other example that does the same.

Yeah, finding examples for these is not easy :sweat_smile: 
I found docs for that in [extending-the-rest-api/schema/#primitive-types](https://developer.wordpress.org/rest-api/extending-the-rest-api/schema/#primitive-types) and by testing it here it works :+1: 
I suppose we _could_ add a [`pattern`](https://developer.wordpress.org/rest-api/extending-the-rest-api/schema/#pattern) for the post-type strings inside the array definition, but that would just complicate things without any apparent benefit other than being stricter on the strings that can be a post-type. Which makes sense... but from what I can tell in this case is of no use (which is why I didn't add a `pattern`)

**Comment by TimothyBJacobs** [comment]
Thanks for adding those tests! Instead of checking for a count, though, let's check the actual user IDs are the ones that we expect.

If we are going to allow for specifying specific post types, then we don't want to use `pattern` but `enum`. So on `items` set an `enum` of `get_post_types( array( 'show_in_rest' => true ), 'names' )`.

We'll also then need to test to ensure that a user can't pass a post type like `custom_css`.

**Comment by aristath** [comment]
> Thanks for adding those tests! Instead of checking for a count, though, let's check the actual user IDs are the ones that we expect.
> 
> If we are going to allow for specifying specific post types, then we don't want to use `pattern` but `enum`. So on `items` set an `enum` of `get_post_types( array( 'show_in_rest' => true ), 'names' )`.
> 
> We'll also then need to test to ensure that a user can't pass a post type like `custom_css`.

Thank you for the suggestions @TimothyBJacobs :+1: 
All done.

**Review by TimothyBJacobs — APPROVED** [review]

**Pull request merged**
